hbase-server/src/main/java/org/apache/hadoop/hbase/security/provider/SaslServerAuthenticationProviders.java

```diff
@@ -0,0 +1,179 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.security.provider;
+
+import java.util.HashMap;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@InterfaceAudience.Private
+public final class SaslServerAuthenticationProviders {
+  private static final Logger LOG = LoggerFactory.getLogger(
+      SaslClientAuthenticationProviders.class);
+
+  public static final String EXTRA_PROVIDERS_KEY = "hbase.server.sasl.provider.extras";
+  private static final AtomicReference<SaslServerAuthenticationProviders> holder =
+      new AtomicReference<>();
+
+  private final Configuration conf;
+  private final HashMap<Byte, SaslServerAuthenticationProvider> providers;
+
+  private SaslServerAuthenticationProviders(Configuration conf,
+      HashMap<Byte, SaslServerAuthenticationProvider> providers) {
+    this.conf = conf;
+    this.providers = providers;
+  }
+
+  /**
+   * Returns the number of registered providers.
+   */
+  public int getNumRegisteredProviders() {
+    return providers.size();
+  }
+
+  /**
+   * Returns a singleton instance of {@link SaslServerAuthenticationProviders}.
+   */
+  public static SaslServerAuthenticationProviders getInstance(Configuration conf) {
+    SaslServerAuthenticationProviders providers = holder.get();
+    if (null == providers) {
+      synchronized (holder) {
+        // Someone else beat us here
+        providers = holder.get();
+        if (null != providers) {
+          return providers;
+        }
+
+        providers = createProviders(conf);
+        holder.set(providers);
+      }
+    }
+    return providers;
+  }
```

Comment:
There used to be a problem with this kind of synchronization in the JDK.  I'm not sure if it's still the case, but I would KISS.  Just synchronize the entire method and not have to worry about any race conditions later.

Reply:
We call this method for every RPC received by a server. "KISS" will be directly harmful in this case.

Reply:
The issue I'm aware of with double-checked locking was around checking a variable that was not marked as `volatile`. Because of the `AtomicReference` used here, we should not be subject to this (as that would mean the guarantees made by that class are invalid).